pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,254 @@
+				return Ok(());
+			}
+
+			// If `from` and the caller are different addresses, a successful transfer results
+			// in decreased allowance by `from` to the caller and an `Approval` event with
+			// the new allowance amount is emitted.
+			api::transfer_from(self.id, from, to, value).map_err(PSP22Error::from)?;
+			// Emit events.
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			self.env().emit_event(Approval {
+				owner: from,
+				spender: caller,
+				value: self.allowance(from, caller),
+			});
+			Ok(())
 		}
 
+		/// Allows `spender` to withdraw from the caller's account multiple times, up to
+		/// the total amount of `value`.
 		#[ink(message)]
-		pub fn increase_allowance(
+		fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), PSP22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address, returns success and no events
+			// are emitted.
+			if caller == spender {
+				return Ok(());
+			}
+
+			api::approve(self.id, spender, value).map_err(PSP22Error::from)?;
+			self.env().emit_event(Approval { owner: caller, spender, value });
+			Ok(())
+		}
+
+		/// Increases by `value` the allowance granted to `spender` by the caller.
+		#[ink(message)]
+		fn increase_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::increase_allowance(token, spender, value)
+		) -> Result<(), PSP22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero, returns
+			// success and no events are emitted.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+
+			api::increase_allowance(self.id, spender, value).map_err(PSP22Error::from)?;
+			let allowance = self.allowance(caller, spender);
+			self.env().emit_event(Approval { owner: caller, spender, value: allowance });
+			Ok(())
 		}
 
+		/// Decreases by `value` the allowance granted to `spender` by the caller.
 		#[ink(message)]
-		pub fn decrease_allowance(
+		fn decrease_allowance(
 			&mut self,
-			token: TokenId,
 			spender: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::decrease_allowance(token, spender, value)
+		) -> Result<(), PSP22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address or `value` is zero, returns
+			// success and no events are emitted.
+			if caller == spender || value == 0 {
+				return Ok(());
+			}
+			// Reverts with `InsufficientAllowance` if `spender` and the caller are different
+			// addresses and the `value` exceeds the allowance granted by the caller to
+			// `spender`.
+			let allowance = self.allowance(caller, spender);
+			if allowance < value {
+				return Err(PSP22Error::InsufficientAllowance);
+			}
+
+			api::decrease_allowance(self.id, spender, value).map_err(PSP22Error::from)?;
+			let allowance = self.allowance(caller, spender);
```

Comment:
This is bad, we should query it twice. Too expensive

Reply:
- Reason for this comes from this comment: https://github.com/r0gue-io/pop-node/pull/300#discussion_r1771017099
- Before it was `allowance.saturating_sub`, we need to query before the API call because

```
TODO: `Fungibles::decrease_allowance()` saturating_sub the value, it should `checked_sub` and throw an error instead. Hence, we don't have to handle `InsufficientAllowance` on the contract side.
```

Right now we need that query to throw error before the API call following the standard.

Reply:
Can we not add the value to the previously queried value?

It is not ideal at all but I don't see another way.

Reply:
Sorry I don't understand what you mean. Do you mean should we add the value to the previously queried value?